### ISSUE TITLE
Inference code for multiple exposure images

### ISSUE THREAD

**Issue by dat-lequoc** [LucasKirsten/MobileMEF#1]
Hi @LucasKirsten,

Thank you for the paper and great work. This will be very helpful for my use case.

![image](https://github.com/user-attachments/assets/32838e01-10ce-40b5-b85c-3a55999210fb)

Ground truth : (brighter than the average, edited by human)
![image](https://github.com/user-attachments/assets/ef0251c2-749d-48f0-b780-e7b9ae8f9fbb)

I'm attempting to use the pretrained model on my dataset, which contains multi-exposure images. I have a few questions:

1. In the paper, you mention the model was trained on at least 2 exposures. Given my data, which specific model in `/h5` would you recommend for my use case?

2. My photos are 5312x3552 pixels. Can the model handle this resolution, which is the best suitable GPUs for inference / train?

3. Considering the differences between my data and your training set, do you think I should:
   a) Use the pretrained model as-is
   b) Fine-tune the pretrained model on my dataset
   c) Train a new model from scratch using my edited photo ground truth

Thank you very much for your insight 🤩!

**Comment by LucasKirsten** [comment]
Hi @dat-lequoc, thanks for your interest in MobileMEF!

1. Your images seem similar to those of the SICE dataset, so I would recommend using either sice_ev1.h5 or sice_ev_most.h5. For choosing which to use, I also recommend testing different input setups of EV frames on each of the models to see which outputs suit you better.

2. The models were trained with an input resolution of 4096x2816, so resizing the images to this resolution would give you the best possible results. However, I believe using other resolutions (such as the one of your photos) would result in a negligible loose of quality. I also tested inference on images with 4096x4096 in my laptop NVIDIA RTX A3000 GPU with 6GB and it was capable of handling the images fine.

3. I recommend starting by using a pre-trained model and verifying which would suit you better (as I mentioned in 1.), and, if the results are not good enough, try to fine-tune the model based on a pre-trained checkpoint. I would not recommend training from scratch, since it can take several hours and require a GPU with high memory (40GB plus).

P.S.: for fine-tuning the model, unfortunately, I can not provide the training scripts, since they are proprietary to Motorola. But feel free to ask me about implementation issues you may have when trying to reproduce the paper.

**Comment by dat-lequoc** [comment]
Thank you very much, @LucasKirsten, for your response.

1. Could you please provide a requirements file (e.g., `pip freeze`) so I can easily install the necessary packages with pip? `Conda` isn't supported and can be difficult to set up for free GPU environments like `Google Colab` or `Kaggle`.

2. How can I modify the code to handle 5 images instead of only 2 (under/over)? Will it works?
Currently, I can only use 2 : 
![image](https://github.com/user-attachments/assets/bae19b6f-7e01-4265-8dba-c2ab4b2a3e77)


Thank you so much again for your help. Have a nice day! 🤗🤗🤗

**Comment by dat-lequoc** [comment]
(1) is solved, I used Runpod to make conda env, then pip freeze , etc ... Here's the working `requirements.txt` : 
```
numpy==1.24.3
pandas==1.5.0
scipy==1.10.1
scikit-learn==1.2.2
tensorflow==2.10.0
keras==2.10.0
tensorboard==2.10.1
tensorboard-data-server==0.6.1
tensorboard-plugin-wit==1.8.1
tensorflow-estimator==2.10.0
Keras-Preprocessing==1.1.2
Pillow==9.4.0
requests==2.31.0
scikit-image==0.19.3
h5py==3.8.0
networkx==3.1
protobuf
grpcio
opencv-python
tqdm
```

**Comment by dat-lequoc** [comment]
The output of the model is great. The inference speed is close to instant, it's incredible ! But for my usecase, it needs more training to adapt with professional editing ground truth.
`since_ev1.h5` : 
![image](https://github.com/user-attachments/assets/4cd9dbf9-6684-4962-9765-eaa46c89c028)
`since_ev_most.h5`:
![image](https://github.com/user-attachments/assets/396fdbc1-e1a1-48e1-8569-b614733cef7c)
`my ground truth`:
![image](https://github.com/user-attachments/assets/2d16e9fe-2c6c-4eb9-be12-ab295e7e3c05)

**Comment by dat-lequoc** [comment]
3. May I also ask how much time it took to train the model with the GPU you described in the paper?
![image](https://github.com/user-attachments/assets/17416444-d2b2-4c3c-a318-37a8521c3b39)

**Comment by LucasKirsten** [comment]
Hi @dat-lequoc.

(1) Great! Installing only the correct version of tensorflow would be enough since there are no major version requirements on the other libs.

(2) The models were trained using only 2 frame inputs, so you could not feed the network more than it at each iteration. The only way for the model to start accepting other input setups would be through retraining. One workaround would be combining the input images 2 by 2 to feed the model and using the outputs as new inputs.

(3) For the SICE dataset it took around 100 epochs/day using the default resolution in the model.py code. Using lower resolutions and training with mixed precision should make it train faster. The main bottleneck is the Gradient Loss, which takes a while to compute. One suggestion is to not train all epochs with Gradient Loss, and start using it only at the final epochs to boost the results quality.

**Comment by dat-lequoc** [comment]
MobileMEF is fast and gives decent results. 

But finally, I decided to go for something like  Conditional Latent Diffusion Model because it's more adapted to the data set. As the expected ground truth isn't only the fusing version of exposure but also professionally edited by photoshop, which makes training difficult. 

Thank you @LucasKirsten again for your help !

**Comment by LucasKirsten** [comment]
> MobileMEF is fast and gives decent results. 
> 
> But finally, I decided to go for something like  Conditional Latent Diffusion Model because it's more adapted to the data set. As the expected ground truth isn't only the fusing version of exposure but also professionally edited by photoshop, which makes training difficult. 
> 
> Thank you @LucasKirsten again for your help !
> 

Hi @dat-lequoc , have you tried using the lenviz_ev_-20_0.h5 weights? I have recently uploaded this new weights file that contains the model trained with an yet to be released large dataset with professional edited GT.

**Comment by dat-lequoc** [comment]
Yeah, I did it but there's no major improvement. I think MobileMEF is great for device, but for my use case, I can afford extensive compute to generate the best image. So Diffusion model is more adapted.
